modules/framework/galasa-parent/dev.galasa.framework.k8s.controller/src/main/java/dev/galasa/framework/k8s/controller/InterruptedRunEventProcessor.java

```diff
@@ -0,0 +1,115 @@
+import dev.galasa.framework.spi.IResultArchiveStore;
+import dev.galasa.framework.spi.IResultArchiveStoreDirectoryService;
+import dev.galasa.framework.spi.IRunResult;
+import dev.galasa.framework.spi.ResultArchiveStoreException;
+import dev.galasa.framework.spi.RunRasAction;
+import dev.galasa.framework.spi.teststructure.TestStructure;
+
+/**
+ * InterruptedRunEventProcessor runs as a thread in the engine controller pod and it maintains a
+ * queue of interrupt events.
+ * 
+ * When the interrupt event queue is not empty, it removes each event from the queue and processes them
+ * by marking interrupted runs as finished in the DSS and updating their RAS records as defined by the
+ * deferred RAS actions within the interrupt event.
+ */
+public class InterruptedRunEventProcessor implements Runnable {
+
+    private Log logger = LogFactory.getLog(getClass());
+
+    private final Queue<RunInterruptEvent> queue;
+    private IFrameworkRuns frameworkRuns;
+    private IResultArchiveStore rasStore;
+
+    public InterruptedRunEventProcessor(Queue<RunInterruptEvent> queue, IFrameworkRuns frameworkRuns, IResultArchiveStore rasStore) {
+        this.queue = queue;
+        this.frameworkRuns = frameworkRuns;
+        this.rasStore = rasStore;
+    }
+
+    @Override
+    public void run() {
+        try {
+            boolean isDone = false;
+
+            logger.debug("Starting scan of interrupt events to process");
+            while (!isDone) {
+
+                RunInterruptEvent interruptEvent = queue.poll();
+                if (interruptEvent == null) {
+                    isDone = true;
+                } else {
+                    markRunFinishedInDss(interruptEvent);
+                    processRasActions(interruptEvent);
+                }
+            }
+            logger.debug("Finished scan of interrupt events to process");
+        } catch (Exception ex) {
+            logger.warn("Exception caught and ignored in InterruptRunEventProcessor", ex);
+        }
+    }
+
+    private void markRunFinishedInDss(RunInterruptEvent interruptEvent) throws DynamicStatusStoreException {
+        String runName = interruptEvent.getRunName();
+        String interruptReason = interruptEvent.getInterruptReason();
+
+        logger.info("Marking run '" + runName + "' as finished in the DSS");
+        frameworkRuns.markRunFinished(runName, interruptReason);
+        logger.info("Marked run '" + runName + "' as finished in the DSS OK");
+    }
+
+    private void processRasActions(RunInterruptEvent interruptEvent) throws ResultArchiveStoreException {
+        List<RunRasAction> rasActions = interruptEvent.getRasActions();
+        String runName = interruptEvent.getRunName();
+
+        logger.info("Processing RAS actions for run '" + runName + "'");
+
+        for (RunRasAction rasAction : rasActions) {
+            String runId = rasAction.getRunId();
+            TestStructure testStructure = getRunTestStructure(runId);
```

Comment:
Subclass TestStructure to be a TestStructureDocument, to contain the _id and the _rev so it can be passed around rather than loaded twice.

Reply:
The TestStructureCouchdb type already exists in the extensions module so I've changed the `updateTestStructure` code to use that in https://github.com/galasa-dev/galasa/pull/273/commits/9268879a4e549b67260cb2fef9cb163d6c0ac963